### ISSUE TITLE
Picker only opens when clicking on the arrow, not anywhere else [Mac]

### ISSUE THREAD

**Issue by j4m3z0r** [xamarin/Xamarin.Forms#1610]
### Description

I have added a Picker to my Xamarin.Forms Mac application, and the only way to get it to open is to click on the small blue square at the right end of it. Clicking on the white region that contains the current selection will not open the Picker. I also can't see a method to have it open programmatically, so I can't solve this by triggering that method on a GestureRecognizer

### Steps to Reproduce

1. Create a Xamarin.Forms Mac application with a Picker
2. Click the white part of the picker where the text goes (not the blue box with the arrow)
3. Observe that the picker does not drop down to reveal its contents.

### Expected Behavior

Clicking anywhere on the picker should open it, so that it behaves the same as other Mac applications (the "Rotation" option in the display preferences on a Mac is a good example).

### Actual Behavior

Clicking anywhere but the blue box does nothing.

### Basic Information

- Version with issue: Xamarin.Forms.2.5.0.121934
- Last known good version: n/a
- IDE: Rider
- Platform Target Frameworks: 
  - Mac: Xamarin.Mac == 4.0.0.214.

**Comment by BradChase2011** [comment]
My understanding is that is the correct native way in macOS.  I thought about the ability to click into the white as well but the control also can be editable.  I know that is not supported under Xamarin.Forms but that is my understanding of why Apple does it that way.

**Comment by j4m3z0r** [comment]
So I wasn't able to find any examples of Apple applications that exhibit that behavior. It's true that if the drop-down is editable then clicking on the edit box won't drop down, but all the non-editable drop-downs I found could be activated by clicking anywhere inside them.

Apple's [Human Interface Guidelines](https://developer.apple.com/library/content/documentation//UserExperience/Conceptual/OSXHIGuidelines/ControlswithMenus.html#//apple_ref/doc/uid/20000957-CH100-SW1) also suggest that you should be able to click anywhere. From the section "Command Pop-Down Menu":

"""
A command pop-down menu:
[...]
* Opens when users click anywhere in the control.
"""

The other types of drop-downs aren't quite so explicit. eg: Pop-*Up* Menu says "A pop-up menu behaves like other menus: Users press to open the menu and then drag to choose an item", but that still reads to me like I should be able to click anywhere, and the example they cite has this behavior (click anywhere opens it, not just click on the arrow).

At any rate, the existing implementation is pretty hard to use -- I certainly don't expect to have to be so precise with my mouse clicks. As a minimum it would be great to offer the option to have it mimic the behavior of all the non-editable drop-downs in the System Preferences app.

**Comment by BradChase2011** [comment]
Ahh ok, that's where the discrepancy is.  The control used is the NSCombobox not the pop-up.  So I think your asking if it can be switched over?  Otherwise custom code will need to be put in to have what your asking done to the combo.

**Comment by BradChase2011** [comment]
Just to note "Don’t extend the right edge of the list beyond the right edge of the arrow box" annoys the crap out of me currently.  So I would second switching out the control.  On the other side, are there performance hits or issues with scrolling if the list is too long?

**Comment by j4m3z0r** [comment]
Ah, I see. So yeah, I'd argue that NSComboBox isn't appropriate here, since my understanding is that the Picker class doesn't expose a way to edit the text in the selected item.

I'm not an expert at AppKit, but it looks like NSPopUpButton is probably the right option.

**Comment by j4m3z0r** [comment]
WRT scalability, I note that NSComboBox uses a dataSource, and NSPopUpButton does not. NSPopUpButton seems to defer all that to NSMenu. So I suppose there's some possibility that it may not scale as well, but I'm yet to encounter a prohibitively slow drop-down menu on my Mac. It seems like an acceptable trade-off for getting a UI that is consistent with other Mac apps.

**Comment by BradChase2011** [comment]
I have to check in a bunch of fixes for macOS still.  Should have time later this week.  I can fix up a renderer for it if you want to check that out in the interim and maybe we can get some feedback on what to do going forward with it.

**Comment by j4m3z0r** [comment]
If it's not too complicated for me to try out (eg: a myget link or something), I'm happy to give it a go and let you know what I think. LMK, and thanks for looking into this.

**Comment by BradChase2011** [comment]
@j4m3z0r 

OK, I made one with the NSPopupButton for ya.  Not gunna lie, I like it quite a bit.  Matter of fact I am going to use it now.  I havent noticed much of a performance drop but on the other hand I havent put it through the perf tools yet and I use a full screen modal picker for large lists including a search and autocomplete feature.  It is MUCH more usable to say the least.  I am definitely thinking of putting up a PR with it, after its refined a bit.  Give it a try and tell me what you think with your application.  I would like to know the usability findings.

https://gist.github.com/BradChase2011/f5a6645d5ecdeba294a70b08cd065e3f

**Comment by BradChase2011** [comment]
To follow up the reality is anything over 30 already starts to degrade performance at a pretty high rate.  Android is probably the worst offender I have seen.  We actually throw an exception at runtime so our developers and testers know the wrong control was used.  So it might fit with this control better.  It does potentially break an apple design rule, but on the other hand I haven't seen them follow them either.

**Comment by j4m3z0r** [comment]
Hey, just wanted to flag that I saw your message but need to get through a few other things on my TODO list before I'll be able to try this out. I take it to use this I need to checkout XF, substitute your gist for the corresponding file and build?

**Comment by BradChase2011** [comment]
@j4m3z0r nope all's you have to do is include it in your project.  As long as the attribute is included at the top, it will be used instead of the one in XF.  So just drop the file in and use :).

**Comment by j4m3z0r** [comment]
@BradChase2011 Ha, that's a neat trick! I just tried it out in my application and this is about a million times better! The drop-downs I have in my application are all < 10 options, and at that scale there's no issue with performance. I haven't tried it with substantially more, but I also probably wouldn't use a drop-down if I needed to select from substantially more options than that.

To be certain, I just had another look at Apple's HIG. The distinction they draw between pop-down and pop-up seems to be whether the items being selected are actions or not. AFAIK this is not a distinction that other platforms make, but as a general rule I think I'd expect this to be used more frequently to select "things" (eg: colors) than actions, so I think your choice of NSPopupButton is probably the most appropriate.

I think this is a winner. You get my vote for merging into master.
